File: python/sort/quick.py
def quickSort(arr, random=False):
    def merge(left, right, mid):
        # left.extend(mid)
        # left.extend(right)
        left[len(left):] = mid
        left[len(left):] = right
        return left

    length = len(arr)

    if length < 2:
        return arr

    if random:
        start = arr[0]
        mid = arr[int(length / 2)]
        end = arr[length - 1]
        if start > mid:
            if start > end:
                pivot = mid if mid > end else end
            else:
                pivot = start
        elif start <= mid:
            if start < end:
                pivot = end if mid > end else mid
            else:
                pivot = start
    else:
        pivot = arr[0]

    leftArr = []
    rightArr = []
    midArr = []
    for i in range(0, length):
        if arr[i] < pivot:
            leftArr.append(arr[i])
        if arr[i] > pivot:
            rightArr.append(arr[i])
        elif arr[i] == pivot:
            midArr.append(arr[i])

    return merge(quickSort(leftArr), quickSort(rightArr), midArr)

File: js/sort/quick.js
/**
 * 快排
 */
var quickSort = function(arr, random) {
  function merge(left, right, mid) {
    return left.concat(mid).concat(right)
  }

  if(arr.length < 2) {
    return arr
  }

  /**
   * 枢纽元
   */
  var pivot
  if(random) {
    var pivotList = [arr[0], arr[parseInt(arr.length / 2)], arr[arr.length - 1]]
      // pivotList.sort(function(a, b) {
      //   return a - b
      // })
    if(pivotList[0] > pivotList[1]) {
      if(pivotList[0] > pivotList[2]) {
        pivot = pivotList[1] > pivotList[2] ? pivotList[1] : pivotList[2]
      } else {
        pivot = pivotList[0]
      }
    } else {
      if(pivotList[0] < pivotList[2]) {
        pivot = pivotList[1] > pivotList[2] ? pivotList[2] : pivotList[1]
      } else {
        pivot = pivot[0]
      }
    }
  } else {
    pivot = arr[0]
  }


  var leftArr = []
  var rightArr = []
  var midArr = []

  for(var i = 0; i < arr.length; i++) {
    if(arr[i] > pivot) {
      rightArr.push(arr[i])
    } else if(arr[i] == pivot) {
      midArr.push(arr[i])
    } else {
      leftArr.push(arr[i])
    }
  }

  return merge(quickSort(leftArr), quickSort(rightArr), midArr)

}

module.exports = quickSort
